props/blaster.h

```diff
@@ -0,0 +1,192 @@
+  };
+
+  NextAction next_action_ = NEXT_ACTION_NOTHING;
+  uint32_t time_base_;
+  uint32_t next_event_time_;
+  
+  void SetNextAction(NextAction what, uint32_t when) {
+    time_base_ = millis();
+    next_event_time_ = when;
+    next_action_ = what;
+  }
+
+  void SetNextActionF(NextAction what, float when) {
+    SetNextAction(what, when * 1000);
+  }
+  
+  void PollNextAction() {
+    if (millis() - time_base_ > next_event_time_) {
+      switch (next_action_) {
+        case NEXT_ACTION_NOTHING:
+          break;
+        case NEXT_ACTION_ARM:
+          armed_ = true;
+          // TODO: Should we have separate ARMING and ARMED states?
+          break;
+        case NEXT_ACTION_BLOW:
+          Off(OFF_BLAST);
+          break;
+      }
+      next_action_ = NEXT_ACTION_NOTHING;
+    }
+  }
+  
+  void beginArm() {
+    SaberBase::SetLockup(SaberBase::LOCKUP_ARMED);
+    SaberBase::DoBeginLockup();
+#ifdef ENABLE_AUDIO    
+    float len = hybrid_font.GetCurrentEffectLength();
+#else    
+    float len = 1.6;
+#endif
+    SetNextActionF(NEXT_ACTION_ARM, len);
+  }
+
+  void selfDestruct() {
+    SaberBase::DoEndLockup();
+#ifdef ENABLE_AUDIO    
+    float len = hybrid_font.GetCurrentEffectLength();
+#else    
+    float len = 0.0;
+#endif
+    SaberBase::SetLockup(SaberBase::LOCKUP_NONE);
+    if (armed_) {
+      SetNextActionF(NEXT_ACTION_BLOW, len);
+    } else {
+      SetNextAction(NEXT_ACTION_NOTHING, 0);
+    }
+  }
+
+  void Loop() override {
+    PropBase::Loop();
+    PollNextAction();
+  }
+
+  // Make clash do nothing
+  void Clash(bool stab) override {}
+
+  // Make swings do nothing
+  void DoMotion(const Vec3& motion, bool clear) override { }
+
+  bool Event2(enum BUTTON button, EVENT event, uint32_t modifiers) override {
+    switch (EVENTID(button, event, modifiers)) {
+      case EVENTID(BUTTON_POWER, EVENT_LATCH_ON, MODE_OFF):
+      case EVENTID(BUTTON_AUX, EVENT_LATCH_ON, MODE_OFF):
+      case EVENTID(BUTTON_AUX2, EVENT_LATCH_ON, MODE_OFF):
+      case EVENTID(BUTTON_POWER, EVENT_CLICK_SHORT, MODE_OFF):
+        armed_ = false;
+        SetPower(true);
+        On();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_LATCH_OFF, MODE_ON):
+      case EVENTID(BUTTON_POWER, EVENT_LATCH_OFF, MODE_OFF):
+      case EVENTID(BUTTON_POWER, EVENT_CLICK_LONG, MODE_ON):
+        SetPower(false);
+        Off();
+        return true;
+
+      case EVENTID(BUTTON_AUX, EVENT_DOUBLE_CLICK, MODE_OFF):
+        if (powered_) rotate_presets();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_DOUBLE_CLICK, MODE_ON):
+        StartOrStopTrack();
+        return true;
+
+      case EVENTID(BUTTON_AUX, EVENT_PRESSED, MODE_ON):
+        Fire();
+        return true;
+
+      case EVENTID(BUTTON_AUX, EVENT_RELEASED, MODE_ON):
+        if (blaster_mode == MODE_AUTO) SaberBase::DoEndLockup();
+        return true;
+
+      case EVENTID(BUTTON_AUX2, EVENT_PRESSED, MODE_ON):
+        beginArm();

```

Comment:
detonators and phases behave differently though.
On a detonator, there is a button, and as long as you're holding it, everything is safe.
Once you release a button, a 4-second countdown starts and then BOOM.

For phasers, you can set it for overload, and it starts making a noise, which becomes louder and louder until it explodes, which can take anything from 20 seconds to a few minutes depending on phaser model I think. However, at any time, it is possible to undo the overload and make it not explode.

On the STNG, overload is apparently done by pressing a few buttons on the back of the phaser. But I like the method they use in the ST pilot better, where they twist the barrel to select different modes, and and if you twist it enough (or maybe leave it between modes) the overload begins.

Reply:
Should the self destruct sequence be a mode like kill, stun, auto then maybe?

Reply:
I think that makes sense.

Reply:
Ok. I can do that. 👍

Reply:
So this is the one that is going to require some thought and work.. do we want to merge as-is, or postpone until I rework the entire self-destruct sequence?

Reply:
Either way is fine with me I think.
Probably the best thing to do would be to merge this soon, and put it in 3.x, and if you have the time, or the 3.x release drags out for some reason, you can submit updates. Otherwise it 3.x would have the code the way it is, which I think is ok.